### ISSUE TITLE
[FEA] Use RMM memory pool by default

### ISSUE THREAD

**Issue by shwina** [rapidsai/cudf#12235]
We should move to using an RMM managed memory pool by default.

This was brought up before in https://github.com/rapidsai/cudf/issues/2676. In response to that issue, we implemented `set_allocator`, https://github.com/rapidsai/cudf/pull/2682, but we chose not to enable the RMM pool by default (likely because we didn't want to monopolize GPU memory away from other libraries). 

Since then, CuPy, Numba (and soon [PyTorch](https://github.com/pytorch/pytorch/pull/86786)) all can be configured to use RMM, and therefore share the same memory pool as cuDF.

## Proposal

Concretely, the proposal is that `import cudf` will:

* Set RMM's default memory resource to a [pool memory resource](https://docs.rapids.ai/api/rmm/stable/api.html#rmm.mr.PoolMemoryResource)
* Configure CuPy, Numba, (and PyTorch?) to all use RMM's default memory resource

## What should the initial and maximum pool size be? 

An RMM pool can be configured with an initial and maximum pool size. The pool grows according to an implementation-defined strategy (see [here](https://github.com/rapidsai/rmm/blob/d132e5236b444d2dcdae25e846c4fe4d5651ee79/include/rmm/mr/device/pool_memory_resource.hpp#L246-L249) for the current strategy).

- As we cannot assume that all (or any) GPU memory is available when cuDF is imported, the initial pool size should be 0 bytes.
- The only reasonable maximum pool size I can think of is the maximum available GPU memory. If the pool cannot expand to this size because of allocations made outside of RMM, so be it: we will OOM. 

## What happens if `import cudf` appears in the middle of the program?

All this works well if `import cudf` appears at the beginning of the program, i.e., before any device memory is actually allocated by any library). However, if it appears _after_ some device objects have already been allocated, it can lead to early out-of-memory errors. As an example, consider some code that uses both PyTorch and cuDF in the following way:

```python
import torch
 
# this part of the code uses PyTorch

import cudf

# this part of the code uses cudf
```

Because PyTorch [uses a caching allocator](https://pytorch.org/docs/stable/notes/cuda.html#memory-management), a memory pool already exists by the time we import cuDF. Importing cuDF initializes a second pool that all libraries (including PyTorch) will use going forward. The first pool essentially becomes a kind of dead space: no new device objects are ever allocated within the pool, and no device memory is ever freed from it. 

There's no perfect solution I can think of to this particular problem, but it's probably a good idea to call [`empty_cache()`](https://pytorch.org/docs/stable/generated/torch.cuda.empty_cache.html#torch.cuda.empty_cache) before resetting the PyTorch allocator to minimize the amount of wasted memory.

---

That's just one example of the kind of issues that can arise if `import cudf` appears later. I think it's fine to assume this will be less common than importing it at the beginning.

**Comment by vyasr** [comment]
Some thoughts:
- cuda-python/pynvml have APIs for checking total device memory usage (e.g. `cudaMemGetInfo`). Should we inject some checks at import time to ensure that the entire GPU is available, or maybe print some initial stats on the max pool size based on available GPU memory?
- Can we wrap the allocator in an adapter that checks (on OOM) whether memory has been allocated outside of the pool (perhaps just using a naive metric like `total_gpu_mem - pool_size`) and provides the user a more useful error?
- I'm always wary of making implicit and far-reaching changes like this silently. Should we consider adding some notification on `import cudf` that indicates that such a change is being made? I _also_ don't like libraries that print information (especially warnings) unconditionally on import, so I don't particularly like this suggestion, just raising the possibility. Alternatively it could be set up via logging of some sort. I think we can safely assume that no matter how we broadcast this some users will not read the docs/notification/warning/etc and be surprised at the behavior, but hopefully (and I think it's likely that) more people are happy about the pool usage improving default performance.
- Similar to the PyTorch example where cudf is imported later, could this change potentially break scripts that are using rmm to allocate memory prior to the cudf import? I believe that the answer is no since we made the necessary changes to save an allocator used for any allocation such that `set_allocator` won't make accessing allocations using the prior allocator UB, just want to be sure.

**Comment by jakirkham** [comment]
On 2, there's at least [a blogpost]( https://medium.com/rapids-ai/pytorch-rapids-rmm-maximize-the-memory-efficiency-of-your-workflows-f475107ba4d4 ) now. Maybe we can do more to promote that?

**Comment by VibhuJawa** [comment]
> On 2, there's at least [a blogpost](https://medium.com/rapids-ai/pytorch-rapids-rmm-maximize-the-memory-efficiency-of-your-workflows-f475107ba4d4) now. Maybe we can do more to promote that?

I agree that blog post is what I link so that is very useful indeed. 

I think having a dedicated place in the  cuDF docs for it will be great as a lot of people who use cuDF  would probably not know what RMM Pool is (similar to people using Pytorch not knowing about memory [management](https://pytorch.org/docs/stable/cuda.html#memory-management) ) so we will need a way to link it from cuDF docs too.

**Comment by harrism** [comment]
The first line of the PR description says:

> using an RMM managed memory pool 

Be careful. "managed memory pool" implies a `pool_memory_resource` with a `managed_memory_resource` (aka Unified Memory, `cudaMallocManaged`) as upstream. Are you saying that should be the default? Or do you mean a `pool_memory_resource` with `cuda_memory_resource` (`cudaMalloc`) as upstream?

**Comment by GregoryKimball** [comment]
There is a great discussion in this issue so far, and it (implicitly) has mostly referred to a default _unmanaged_ (device-only) pool. The issue of a default memory resource is coming up again in the context of preventing OOM errors for the single workstation, beginner cuDF user, where a default _managed_ memory resource could be a good solution. 

I'm interested to learn more about the default pool behavior in cupy and pytorch as mentioned [above](https://github.com/rapidsai/cudf/issues/12235#issuecomment-1331229185).

**Comment by harrism** [comment]
Another concern I have in this issue is the recommendation to start with a pool of 0 bytes. With the current pool implementation this will create a lot of fragmentation.

**Comment by vyasr** [comment]
See also #14498

**Comment by vyasr** [comment]
We now enable managed memory pools (`PoolMemoryResource<ManagedMemoryResource>`) by default in both cudf.pandas and cudf-polars. I think that is sufficient to consider this issue closed. I would prefer that we treat those front-ends as opinionated while keeping the cudf classic APIs relatively unopinionated. Over time I think we will be moving more and more usage towards the opinionated front-ends in any case.
